core/server/services/bulk-email/index.js

```diff
@@ -36,33 +36,56 @@ module.exports = {
-            return Promise.map(chunkedRecipients, (toAddresses) => {
-                const recipientVariables = {};
-                toAddresses.forEach((email) => {
-                    recipientVariables[email] = recipientData[email];
-                });
 
-                const messageData = Object.assign({}, message, {
-                    to: toAddresses,
-                    from: fromAddress,
-                    'recipient-variables': recipientVariables
+        const blogTitle = settingsCache.get('title');
+        fromAddress = blogTitle ? `${blogTitle}<${fromAddress}>` : fromAddress;
+
+        const chunkedRecipients = _.chunk(recipients, BATCH_SIZE);
+
+        return Promise.mapSeries(chunkedRecipients, (toAddresses) => {
+            const recipientVariables = {};
+            toAddresses.forEach((email) => {
+                recipientVariables[email] = recipientData[email];
+            });
+
+            const batchData = {
+                to: toAddresses,
+                from: fromAddress,
+                'recipient-variables': recipientVariables
+            };
+
+            const bulkEmailConfig = configService.get('bulkEmail');
+
+            if (bulkEmailConfig && bulkEmailConfig.mailgun && bulkEmailConfig.mailgun.tag) {
+                Object.assign(batchData, {
+                    'o:tag': [bulkEmailConfig.mailgun.tag, 'bulk-email']
                 });
-                const bulkEmailConfig = configService.get('bulkEmail');
+            }
+
+            const messageData = Object.assign({}, message, batchData);
 
-                if (bulkEmailConfig && bulkEmailConfig.mailgun && bulkEmailConfig.mailgun.tag) {
-                    Object.assign(messageData, {
-                        'o:tag': [bulkEmailConfig.mailgun.tag, 'bulk-email']
-                    });
-                }
+            return new Promise((resolve) => {
+                mailgunInstance.messages().send(messageData, (error, body) => {
+                    if (error) {
+                        // NOTE: logging an error here only but actual handling should happen in more sophisticated batch retry handler
+                        // REF: possible mailgun errors https://documentation.mailgun.com/en/latest/api-intro.html#errors
+                        common.logging.error(new common.errors.GhostError({
+                            err: error,
+                            context: 'The bulk email service was unable to send a message, your site will continue to function.',
+                            help: common.i18n.t('errors.services.ping.requestFailed.help', {url: 'https://ghost.org/docs/'})
+                        }));
 
-                return mailgunInstance.messages().send(messageData);
+                        // NOTE: these are generated variables, so can be regenerated when retry is done
+                        const data = _.omit(batchData, ['recipient-variables']);
+                        resolve({
```

Comment:
Maybe we wanna resolve with something like:
```js
resolve(Success(data))
// OR
resolve(Failure(data))
```

Check out the Either monad for inspiration :stuck_out_tongue: